build/build-bundle.js

```diff
@@ -48,143 +57,182 @@ const isLightrider = file => path.basename(file).includes('lightrider');
-    bundle = bundle.require(audit, {expose: audit.replace(corePath, '../')});
-  });
-  gatherers.forEach(gatherer => {
-    bundle = bundle.require(gatherer, {expose: gatherer.replace(driverPath, '../gather/')});
-  });
-
-  // HACK: manually include the lighthouse-plugin-publisher-ads audits.
-  if (isDevtools(entryPath) || isLightrider(entryPath)) {
-    bundle.require('lighthouse-plugin-publisher-ads');
-    pubAdsAudits.forEach(pubAdAudit => {
-      bundle = bundle.require(pubAdAudit);
-    });
+  for (const modulePath of modulesToIgnore) {
+    shimsObj[modulePath] = 'export default {}';
   }
 
-  // browerify's url shim doesn't work with .URL in node_modules,
-  // and within robots-parser, it does `var URL = require('url').URL`, so we expose our own.
-  // @see https://github.com/GoogleChrome/lighthouse/issues/5273
-  const pathToURLShim = require.resolve('../lighthouse-core/lib/url-shim.js');
-  bundle = bundle.require(pathToURLShim, {expose: 'url'});
+  const packageJsonShim = {version: require('../package.json').version};
+  shimsObj[require.resolve('../package.json')] =
+    `export default ${JSON.stringify(packageJsonShim)}`;
 
-  let bundleStream = bundle.bundle();
+  const bundle = await rollup.rollup({
+    input: entryPath,
+    context: 'globalThis',
+    plugins: [
+      replace({
+        delimiters: ['', ''],
+        values: {
+          '/* BUILD_REPLACE_BUNDLED_MODULES */': `[\n${bundledMapEntriesCode},\n]`,
+          '__dirname': (id) => `'${path.relative(LH_ROOT, path.dirname(id))}'`,
+          '__filename': (id) => `'${path.relative(LH_ROOT, id)}'`,
+          // This package exports to default in a way that causes Rollup to get confused,
+          // resulting in MessageFormat being undefined.
+          'require(\'intl-messageformat\').default': 'require(\'intl-messageformat\')',
+        },
+      }),
+      alias({
+        entries: {
+          'debug': require.resolve('debug/src/browser.js'),
+          'lighthouse-logger': require.resolve('../lighthouse-logger/index.js'),
+          'url': require.resolve('../lighthouse-core/lib/url-shim.js'),
+        },
+      }),
+      shim({
+        ...shimsObj,
+        // Allows for plugins to import lighthouse.
+        'lighthouse': `
+          import Audit from '${require.resolve('../lighthouse-core/audits/audit.js')}';
+          export {Audit};
+        `,
+      }),
+      // Currently must run before commonjs (brfs does not support import).
+      // This currenty messes up source maps.
+      rollupBrfs({
+        readFileSyncTransform: minifyFileTransform,
+        global: true,
+        parserOpts: {ecmaVersion: 12, sourceType: 'module'},
+      }),
+      commonjs({
+        // https://github.com/rollup/plugins/issues/922
+        ignoreGlobal: true,
+      }),
+      json(),
+      nodeResolve({preferBuiltins: true}),
+      nodePolyfills(),
+      // Rollup sees the usages of these functions in page functions (ex: see AnchorElements)
```

Comment:
yikes, *this* is going to be really annoying to remember. maybe we should tweak our `evaluate` API to support a keyed-object so we don't have to think about it? 

or better yet always force deps to be function parameters? there are just too many gotchas these days with the `.toString` name reliance

Reply:
I'm for it. https://github.com/GoogleChrome/lighthouse/pull/10781#:~:text=Idea%203,Example

Reply:
IMO if we're going to keep the magically-interspersed Node and `evaluate()` browser code (which #10816 was a culmination of us deciding it was a lot nicer to go that way) we should embrace explicit import of functions and not require everything to live in `page-functions` (a lot of times it's a lot nicer just to have support functions defined locally).

This would allow type checking of functions we currently `@ts-expect-error` and make evaluated code more robust all without much added ceremony. The downside is minifiers are still kind of dumb about keeping consistent minified functions names across modules even after bundling, but there are ways around that or we could just live with somewhat worse minification.

I have an old branch for this I can put up for discussion since I know there's a variety of opinions on each of these points :)

Reply:
(...expanding on the fact that custom page functions outside page-functions.js exists ...)

My linked idea (having a `pageFunctions` namespace) can support custom functions like this:

```js
const pageFunctions = {
  ...require('../../lib/page-functions.js'),
  myFnDeclaredHere,
};

function myFnDeclaredHere() { }

// ...

function getMetaElements() {
  return pageFunctions.getElementsInDocument('head meta').map(el => {
    const meta = /** @type {HTMLMetaElement} */ (el);
    pageFunctions.myFnDeclaredHere(); // also i exist.
    return {
      name: meta.name.toLowerCase(),
      content: meta.content,
      property: meta.attributes.property ? meta.attributes.property.value : undefined,
      httpEquiv: meta.httpEquiv ? meta.httpEquiv.toLowerCase() : undefined,
      charset: meta.attributes.charset ? meta.attributes.charset.value : undefined,
    };
  });
}

const code = pageFunctions.createEvalCode(getMetaElements, {
      deps: [
        pageFunctions.getElementsInDocument,
		pageFunctions.myFnDeclaredHere,
      ],
    });

// `pageFunctions.createEvalCode` would inject a var `pageFunctions` with all values passed to deps, using the `.name` as keys
```

in theory, that should keep the mangled names the same

Reply:
#12795 has not been resolved–is this blocking for this PR?